### ISSUE TITLE
Empire wishlist after WRCCDC

### ISSUE THREAD

**Issue by Viss** [EmpireProject/Empire#476]
## Empire Version
2.0_beta

I hope I'm not stepping on toes by submitting things in this way. I wanted to use this as an opportunity to see if other people who do redteam/attack stuff had the same sort of wants that I do. I was recently fortunate enough to compete on the red team at WRCCDC and I used empire almost exclusively, and I came away with a few items that fit on a "dang, it would be really nice to have these" sort of list, and I figure I'd at least document them so I'm not operating in a vacuum.

- Ability to migrate, vs ps injection
   - having the ability to "migrate 12345" would make workflow considerably faster when seconds/minutes matter. Having 25 "powershell.exe" processes running is quite obvious. 

- Spinning off binaries
   - much of the ccdc workflow involves persistence, and while empire is REALLY GOOD AT THAT for us, the minute the kids learn that removing network access from powershell is a thing every single one of our shells die. I'm wondering if there's a way to continue to maintain access to an environment once powershell is no longer allowed to do network communications? I think a binary may fix this, depending on how this is implemented.

- Displaying/modifying the contents of autorun. 
   - the current functionality of autorun works fairly well, until such time as you forget what the autorun actually is. A shell comes in, "job asdf started" is displayed and ... "something happens"? Some idea of what job it is that is actually being run would be very helpful, or some way to manage the autorun put in place would be very helpful

- DNS comms c2 option
  - basically incorporating dnscat in some way, so that the c2 back to the empire infrastructure can be done over dns, vs http/s

Any or all of these would produce much glee in the CCDC ranks :)

**Comment by xorrior** [comment]
Migrate vs psinject 
We're open to investigating this functionality but could you provide some more information on how this is beneficial or different over injecting into another process and killing the old process.

Binaries
Adding the ability to generate exe's could definitely be added down the road. Managed vs unmanaged code will dictate how long this would take. I would suggest using a small c# binary to execute empire stagers in the mean time. 

Autorun
Definitely can be done in the near future.

Dns comms
Already being worked on.

**Comment by Viss** [comment]
The intention with "migrate" would basically be to maintain the current shell, versus starting a new one. The purpose is that you get different abilities when inside of different processes. If you're not in a users context, you can't take screenshots. If you're not in a system/escalated context, you can't mimikatz - that sort of thing.

Being able to keep the same shell will save time, save keystrokes, and prevent new shells from having to be staged. It would make empire quieter on the wire. Currently empire doesn't produce "stageless shells", which is something I'm somewhat spoiled by from metasploit. It's something that makes life REALLY EASY in that I can

- Create a stageless payload (meterpreter/reverse_tcp_rc4)
- Encode it with veil or BDP or ebowla
- Place it on a machine somewhere, and have it run out of autorun or a registry call

This means the only network communications that happen are purely the c2 comms, and not re-staging a whole shell every time. This is why I was asking for the ability to export binaries, because it would make comms quieter, presuming the binary was stageless. 

Also, because there's no tab completion on OSX, and having to continually type out every shells uniquely generated name every single time I want to try to migrate to another process makes workflow drag really slowly.

**Comment by xorrior** [comment]
Closing.
